Add ContractAddressID to cat storage tables

paytonrules opened the pull request:
A few things that make this larger than might be expected.

* The data_generator needed a hack because CatIlkChop and CatIlkLump now require an addressID, but the rest of the ilk tables don't.
* InsertFieldWithIlkAndAddress was added to shared/utlities, because not every InsertFieldWithIlk needs an address
* I changed MappingResWithAddress to take an int64. Since this is just in test code, it should be fine, but it caused a large number of cascading changes I didn't anticipate.

paytonrules committed:
Add address_id to cat_live storage type.

This also updates its component test. Updating the repository test is
going to cause all the cat storage types to be changed, so that will
happen on the next commit.

paytonrules committed:
Add a repository test

I don't know who made the storage_repository_behaviors, but in the words
of Liam Neeson, I will look for them, I will find them, and I will....

paytonrules committed:
Add address_id to Cat Vat

paytonrules committed:
Add address_id to cat_vow storage

paytonrules committed:
Add address_id to cat_ilk tables

This caused a huge, and unfortunate, series of changes.

- Adding to cat_ilk_flip forced adding to cat_ilk_chop and cat_ilk_lump
- The data_generator needed a hack because CatIlkChop and CatIlkLump now
require an addressID
- InsertFieldWithIlkAndAddress was added to shared/utlities, because not
every InsertFieldWithIlk needs an address
- I changed MappingResWithAddress to take an int64. Since this is just
in test code, it should be fine, but it caused a large number of
cascading changes I didn't anticipate.

paytonrules committed:
Add tests for Chop and Flop

Unit tests in the repository.

paytonrules committed:
Add two more component tests

These were missing, because the original changes were driven by the compiler.

rmulhol commented on data_generator/data_generator.go:
```diff
@@ -397,17 +398,30 @@ func (state *GeneratorState) insertInitialIlkData(ilkId int64) error {
+		// HACK because the cat queries now take an address_id. If the other storage queries start taking an
+		// address_id, remove this hack
+		var err error
+		if intInsertSql == cat.InsertCatIlkChopQuery || intInsertSql == cat.InsertCatIlkLumpQuery {
```
I guess we never deal with `cat_ilk_flip` in here?

paytonrules replied:
Oddly no, but I guess since it's test data.

rmulhol commented on transformers/component_tests/cat/configured_transformer_test.go:
```diff
@@ -136,22 +145,26 @@ var _ = Describe("Executing the transformer", func() {
+			ilkID              int64
+			ilkErr             error
+			contractAddressID  int64
+			contractAddressErr error
```
nbd but I feel like `ilkErr` and `contractAddressErr` could be declared within the `BeforeEach` so that they're not in scope for subsequent `It` statements - seems like we only care about them in the expectations on lines 165 + 167

paytonrules replied:
👍

rmulhol commented on transformers/component_tests/flap/configured_transformer_test.go:
```diff
@@ -189,7 +189,7 @@ var _ = Describe("Executing the flap transformer", func() {
 			err := db.Get(&wardsResult, `SELECT diff_id, header_id, address_id, usr AS key, wards.wards AS value FROM maker.wards`)
 			Expect(err).NotTo(HaveOccurred())
-			Expect(wardsResult.AddressID).To(Equal(strconv.FormatInt(flapAddressID, 10)))
+			Expect(wardsResult.AddressID).To(Equal(flapAddressID))
```
😤

rmulhol commented on transformers/shared/utilities.go:
```diff
@@ -124,6 +124,37 @@ func InsertFieldWithIlk(diffID, headerID int64, ilk, variableName, query, value
+		return fmt.Errorf("error beginning transaction: %w", txErr)
+	}
+
+	addressID, addressErr := GetOrCreateAddress(address, db)
```
Maybe worth using `GetOrCreateAddressInTransaction` here, and rolling back on failure?

paytonrules replied:
👍  - I'm also assuming that eventually these two functions will coalesce into one. Is that fair?

paytonrules replied:
Did not know about that but it makes a lot of sense, since this is in a transaction.

paytonrules replied:
I ended up just passing the addressID here, because we didn't use this method as often as I originally thought. That meant it used the cached version in the cat storage repository.

rmulhol commented on transformers/storage/cat/repository.go:
```diff
@@ -51,23 +51,38 @@ func (repository *StorageRepository) SetDB(db *postgres.DB) {
 
 func (repository *StorageRepository) insertLive(diffID, headerID int64, live string) error {
-	_, err := repository.db.Exec(insertCatLiveQuery, diffID, headerID, live)
+	addressID, addressErr := shared.GetOrCreateAddress(repository.ContractAddress, repository.db)
```
Given the number of times that these inserts can happen, I think it might be cool to have a singleton-ish helper function that memoizes the addressID the first time it's called - so that we don't need to go to the DB for it every time

rmulhol replied:
fwiw we already do what I'm suggesting here: https://github.com/makerdao/vdb-mcd-transformers/blob/staging/transformers/storage/median/repository.go#L292

paytonrules replied:
That makes sense to me - originally I had this passing around the ID but I didn't care for that.

rmulhol commented on transformers/storage/cat/repository.go:
```diff
@@ -51,23 +51,38 @@ func (repository *StorageRepository) SetDB(db *postgres.DB) {
-	_, err := repository.db.Exec(insertCatLiveQuery, diffID, headerID, live)
+	addressID, addressErr := shared.GetOrCreateAddress(repository.ContractAddress, repository.db)
+	if addressErr != nil {
+		return fmt.Errorf("Could not retrieve address id for %s, error: %w", repository.ContractAddress, addressErr)
```
incredibly minor but I think the linter complains about capitalization in error strings

paytonrules replied:
Oh bite me linter. why can no haz capitals?

rmulhol commented on transformers/component_tests/cat/configured_transformer_test.go:
```diff
@@ -136,22 +145,26 @@ var _ = Describe("Executing the transformer", func() {
 			Expect(err).NotTo(HaveOccurred())
-			Expect(wardsResult.AddressID).To(Equal(strconv.FormatInt(catAddressID, 10)))
+			Expect(wardsResult.AddressID).To(Equal(catAddressID))
 			test_helpers.AssertMapping(wardsResult.MappingRes, wardsDiff.ID, header.Id, strconv.FormatInt(userAddressID, 10), "1")
```
Definitely not a blocker but could be cool to introduce a `AssertMappingWithAddress` helper so that we don't need to remember to always include a separate expectation for the `AddressID` if we're doing this more often

paytonrules replied:
Heh - I flip/flopped on that. I can't remember why I stayed with this version.

rmulhol reviewed (approved):
LGTM! A few questions, nothing blocking

paytonrules committed:
PR feedback

- Cache the contractAddressID instead of querying it all the time.
- Pass the address id to the utility function instead of the address so
we don't need to look it up yet again.
- Cleanup the configured transformer test a little, moving variables
that aren't used repeatedly into the BeforeEach.

paytonrules committed:
Replace AssertMapping with AssertMappingWithAddress

Where appropriate - this means you won't forget the addressID.
